<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Pandas (EDA y Unión de Datos)

## Los datos

Para este laboratorio, trabajarás con los siguientes conjuntos de datos:

- **`netflix_originals.csv`**: Contiene información sobre producciones originales de Netflix. Las columnas que nos encontraremos en este conjunto de datos son: 

   - `Title`: Contiene el nombre del título de la producción original de Netflix.

   - `Genre`: Especifica el género o combinación de géneros de la producción.

   - `Premiere`: Indica la fecha de estreno de la producción en Netflix.

   - `Runtime`: Contiene la duración de la producción en minutos.

   - `IMDB Score`: Representa la calificación promedio que la producción ha recibido en la plataforma IMDB.

   - `Language`: Especifica el idioma principal o los idiomas en los que se presenta la producción.

- **`netflix_titles.csv`**: Contiene información sobre una variedad de títulos disponibles en la plataforma Netflix, incluyendo tanto películas como series de televisión. Las columnas que nos encontraremos en este conjunto de datos son:

   - `show_id`: Identificador único para cada título en el dataset.

   - `type`: Especifica el tipo de contenido, ya sea "Movie" (película) o "TV Show" (serie de televisión).

   - `title`: Contiene el nombre del título de la producción.

   - `director`: Muestra el nombre del director del título. Puede estar vacío en algunos casos.

   - `cast`: Lista de actores y actrices que participaron en el título. También puede estar vacío en algunos casos.

   - `country`: País o países donde se produjo el título. Esta columna también puede tener valores faltantes.

   - `date_added`: Fecha en la que el título fue añadido al catálogo de Netflix.

   - `release_year`: Año en que el título fue lanzado o producido.

   - `rating`: Clasificación del contenido en términos de audiencia (por ejemplo, PG-13, TV-MA).

   - `duration`: Duración del título. Para películas, se expresa en minutos (e.g., "90 min") y para series en temporadas (e.g., "2 Seasons").

   - `listed_in`: Categorías o géneros bajo los cuales el título está clasificado (por ejemplo, "Documentaries", "TV Dramas").

   - `description`: Breve sinopsis o descripción del contenido del título.

A lo largo de los laboratorios de Pandas, trabajarás con estos conjuntos de datos (o con los derivados del Laboratorio), algunas de las actividades que realizaremos son:

- Exploración de datos y análisis exploratorio (EDA) para familiarizarse con los conjuntos de datos.

- Realización de uniones entre conjuntos de datos, aplicando diferentes técnicas de *merging* y analizando la relevancia de cada método.

- Uso de funciones de selección y filtrado (`loc`, `iloc`) para extraer información clave.

- Desarrollo de análisis descriptivos y visualización de patrones para comprender mejor las decisiones estratégicas de la plataforma.

----

## Objetivo General de la Serie de Laboratorios



1. **Unión de `netflix_titles.csv` con `netflix_originals.csv`:**

   - **Análisis de contenido original:**  Identificarás los títulos que son producciones originales de Netflix y los compararás con los títulos adquiridos. Este análisis te permitirá:

     - Comparar la diversidad de géneros entre los títulos originales y no originales.

     - Evaluar si ciertos géneros o tipos de contenido (como documentales o series) son más frecuentes en las producciones originales.

     - Investigar si la producción original se concentra en determinados países o idiomas.


   - **Evaluación de contenido:** En esta etapa, explorarás si las producciones originales de Netflix tienden a recibir mejores o peores evaluaciones en comparación con el contenido adquirido. Algunas preguntas que se buscarán responder incluyen:

     - ¿Los títulos originales tienen una calificación promedio superior en plataformas como IMDB?

     - ¿Son las producciones originales más consistentes en términos de calidad percibida?

     - ¿Existen diferencias en la popularidad de los títulos originales según su género o país de origen?


   - **Identificación de tendencias:**  Por último, este análisis te permitirá explorar cómo ha evolucionado la estrategia de contenido de Netflix a lo largo del tiempo. Algunos puntos clave a investigar serán:

     - ¿Cómo ha crecido la producción original en comparación con la adquisición de contenido?

     - ¿Qué géneros o tipos de contenido han recibido mayor inversión en los últimos años?
     
     - ¿Existen patrones en los países o idiomas de las producciones originales, especialmente en relación con la expansión global de Netflix?


----


## Ejercicio: Análisis y Enriquecimiento del Catálogo de Netflix

### Parte 1: Análisis Exploratorio de Datos (EDA)

1. **Carga de los datos:**

   - Deberás cargar los dos conjuntos de datos que te hemos proporcionado para realizar este Laboratorio:

     - `netflix_originals.csv`

     - `netflix_titles.csv`

   Deberás revisar la estructura de cada conjunto de datos utilizando los métodos aprendidos durante la lección.

2. **Análisis de columnas y consistencia:**

   - Deberás hacer un análisis exploratorio de cada uno de los conjuntos de datos. Será obligatorio que incluyas las conclusiones generales obtenidas de cada uno de los análisis exploratorios en el Jupyter Notebook de tus soluciones.

   - Explorar qué columnas contiene cada conjunto de datos y cuáles son sus tipos de datos. Identificar si hay valores nulos o duplicados.

   - Algunas preguntas que nos podemos plantear en esta fase son (recuerda que esto es solo un ejemplo, y que podemos explorar todo lo que queramos/necesitemos para entender lo mejor posible los conjuntos de datos):

      - ¿Cuáles son los tipos de datos de cada columna? Nos puede ayudar a detectar si algún tipo de dato está incorrectamente asignado, como fechas almacenadas como cadenas de texto.

      - ¿Existen valores nulos? ¿En qué columnas están? Este paso es crucial para decidir si deben ser eliminados, rellenados o ignorados.

      - ¿Cuántos títulos únicos hay en cada conjunto de datos? Nos puede permitir verificar si existen títulos duplicados que puedan distorsionar el análisis.

      - ¿Qué géneros tenemos en el conjunto de datos? Esto nos puede ayudar a identificar los géneros más populares o incluso a identificar una posible limpieza de dicha columna para poder extraer conclusiones más claras en nuestro análisis.

### Parte 2: Unión de Datos

1. **Realización de la unión:**

   - Deberás realizar la unión que consideres más apropiada entre `netflix_titles.csv` y `netflix_originals.csv`. Además, deberás justificar el tipo de unión que has realizado.

   - Esto nos ayudará a contestar preguntas más adelente del tipo:

     - ¿Cuántos títulos en `netflix_titles.csv` son producciones originales de Netflix?

     - ¿Hay diferencias significativas en la puntuación de IMDB entre títulos originales y no originales?



# Parte 1

. **Carga de los datos:**

   - Deberás cargar los dos conjuntos de datos que te hemos proporcionado para realizar este Laboratorio:

     - `netflix_originals.csv`

     - `netflix_titles.csv`

   Deberás revisar la estructura de cada conjunto de datos utilizando los métodos aprendidos durante la lección.



In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None) #Ver todo
netflix_ogs = pd.read_csv("datos/netflix_originals.csv")
netflix_titles = pd.read_csv("datos/netflix_titles.csv")

2. **Análisis de columnas y consistencia:**

   - Deberás hacer un análisis exploratorio de cada uno de los conjuntos de datos. Será obligatorio que incluyas las conclusiones generales obtenidas de cada uno de los análisis exploratorios en el Jupyter Notebook de tus soluciones.

   - Explorar qué columnas contiene cada conjunto de datos y cuáles son sus tipos de datos. Identificar si hay valores nulos o duplicados.

   - Algunas preguntas que nos podemos plantear en esta fase son (recuerda que esto es solo un ejemplo, y que podemos explorar todo lo que queramos/necesitemos para entender lo mejor posible los conjuntos de datos):

   - ¿Cuáles son los tipos de datos de cada columna? Nos puede ayudar a detectar si algún tipo de dato está incorrectamente asignado, como fechas almacenadas como cadenas de texto.

- ¿Existen valores nulos? ¿En qué columnas están? Este paso es crucial para decidir si deben ser eliminados, rellenados o ignorados.


In [3]:
netflix_ogs.head(15)

Unnamed: 0                            Title                    Genre  \
0            0                  Enter the Anime              Documentary   
1            1                      Dark Forces                 Thriller   
2            2                          The App    Science fiction/Drama   
3            3                   The Open House          Horror thriller   
4            4                      Kaali Khuhi                  Mystery   
5            5                            Drive                   Action   
6            6                Leyla Everlasting                   Comedy   
7            7  The Last Days of American Crime      Heist film/Thriller   
8            8                          Paradox  Musical/Western/Fantasy   
9            9               Sardar Ka Grandson                   Comedy   
10          11                         The Call                    Drama   
11          12                          Whipped          Romantic comedy   
12          13               All Because of You            Action comedy   
13          14                            Mercy                 Thriller   
14          15                   After the Raid              Documentary   

              Premiere  Runtime  IMDB Score          Language  
0       August 5, 2019       58         2.5  English/Japanese  
1      August 21, 2020       81         2.6           Spanish  
2    December 26, 2019       79         2.6           Italian  
3     January 19, 2018       94         3.2           English  
4     October 30, 2020       90         3.4             Hindi  
5     November 1, 2019      147         3.5             Hindi  
6     December 4, 2020      112         3.7           Turkish  
7         June 5, 2020      149         3.7           English  
8       March 23, 2018       73         3.9           English  
9         May 18, 2021      139         4.1             Hindi  
10   November 27, 2020      112         4.1            Korean  
11  September 18, 2020       97         4.1        Indonesian  
12     October 1, 2020      101         4.2             Malay  
13   November 22, 2016       90         4.2           English  
14   December 19, 2019       25         4.3           Spanish

In [4]:
columnas_og_titles = netflix_ogs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 513 entries, 0 to 512
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  513 non-null    int64  
 1   Title       513 non-null    object 
 2   Genre       513 non-null    object 
 3   Premiere    513 non-null    object 
 4   Runtime     513 non-null    int64  
 5   IMDB Score  513 non-null    float64
 6   Language    513 non-null    object 
dtypes: float64(1), int64(2), object(4)
memory usage: 28.2+ KB


In [5]:
netflix_titles.head()

Unnamed: 0 show_id     type                  title         director  \
0           0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1           1      s2  TV Show          Blood & Water              NaN   
2           2      s3  TV Show              Ganglands  Julien Leclercq   
3           3      s4  TV Show  Jailbirds New Orleans              NaN   
4           4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA        NaN   
3  September 24, 2021          2021  TV-MA        NaN   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [6]:
print(f"Total Duplicados en show_id:")
print(netflix_titles.duplicated(subset="show_id").sum())

Total Duplicados en show_id:
0


In [7]:
columnas_titles = netflix_titles.info()
print()
print("Valores nulos")
print(netflix_titles.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Unnamed: 0    8807 non-null   int64 
 1   show_id       8807 non-null   object
 2   type          8807 non-null   object
 3   title         8807 non-null   object
 4   director      6173 non-null   object
 5   cast          7982 non-null   object
 6   country       7976 non-null   object
 7   date_added    8797 non-null   object
 8   release_year  8807 non-null   int64 
 9   rating        8803 non-null   object
 10  duration      3994 non-null   object
 11  listed_in     8807 non-null   object
 12  description   8807 non-null   object
dtypes: int64(2), object(11)
memory usage: 894.6+ KB

Valores nulos
Unnamed: 0         0
show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating

In [8]:

titles_tipos_unicos = netflix_titles
titles_tipos_unicos.select_dtypes("object").columns

#Encontrar cuantas veces se repiten los valores
for column in titles_tipos_unicos.select_dtypes("object").columns:
    print("")
    print("------------------")
    print(column.upper())
    print(titles_tipos_unicos[column].value_counts())
    print("------------------\n")



------------------
SHOW_ID
show_id
s8807    1
s1       1
s2       1
s3       1
s4       1
        ..
s12      1
s11      1
s10      1
s9       1
s8       1
Name: count, Length: 8807, dtype: int64
------------------


------------------
TYPE
type
Movie      6131
TV Show    2676
Name: count, dtype: int64
------------------


------------------
TITLE
title
Zubaan                                 1
Dick Johnson Is Dead                   1
Blood & Water                          1
Ganglands                              1
Jailbirds New Orleans                  1
                                      ..
Bangkok Breaking                       1
Vendetta: Truth, Lies and The Mafia    1
The Starling                           1
The Great British Baking Show          1
Sankofa                                1
Name: count, Length: 8807, dtype: int64
------------------


------------------
DIRECTOR
director
Rajiv Chilaka              19
Raúl Campos, Jan Suter     18
Suhas Kadav                16
Marc

In [9]:
#tipos en csv 
print(netflix_ogs["Genre"].value_counts())

Genre
Documentary                             132
Drama                                    73
Comedy                                   42
Romantic comedy                          35
Thriller                                 33
                                       ... 
Action-adventure                          1
Concert Film                              1
Animation / Science Fiction               1
Anthology/Dark comedy                     1
Animation/Christmas/Comedy/Adventure      1
Name: count, Length: 106, dtype: int64


In [10]:
#tipos en csv 2
print(titles_tipos_unicos["type"].value_counts())

type
Movie      6131
TV Show    2676
Name: count, dtype: int64


In [11]:
print(netflix_ogs["Genre"].value_counts())

Genre
Documentary                             132
Drama                                    73
Comedy                                   42
Romantic comedy                          35
Thriller                                 33
                                       ... 
Action-adventure                          1
Concert Film                              1
Animation / Science Fiction               1
Anthology/Dark comedy                     1
Animation/Christmas/Comedy/Adventure      1
Name: count, Length: 106, dtype: int64


In [12]:
print(titles_tipos_unicos["listed_in"].value_counts())

listed_in
Dramas, International Movies                        362
Documentaries                                       359
Stand-Up Comedy                                     334
Comedies, Dramas, International Movies              274
Dramas, Independent Movies, International Movies    252
                                                   ... 
Action & Adventure, Cult Movies                       1
Action & Adventure, Comedies, Music & Musicals        1
Classic Movies, Horror Movies, Thrillers              1
Children & Family Movies, Classic Movies, Dramas      1
Cult Movies, Dramas, Thrillers                        1
Name: count, Length: 514, dtype: int64


In [13]:
print(titles_tipos_unicos["rating"].value_counts())

rating
TV-MA       3207
TV-14       2160
TV-PG        863
R            799
PG-13        490
TV-Y7        334
TV-Y         307
PG           287
TV-G         220
NR            80
G             41
TV-Y7-FV       6
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: count, dtype: int64


In [14]:
#Títulos únicos CSV 1
netflix_ogs["Title"].describe()

count                                          513
unique                                         513
top       David Attenborough: A Life on Our Planet
freq                                             1
Name: Title, dtype: object

In [15]:
#Títulos únicos CSV 2
titles_tipos_unicos["title"].describe()

count       8807
unique      8807
top       Zubaan
freq           1
Name: title, dtype: object

# Conclusión Parte 1
Nota: Estas conclusiones son previas a realizar la unión

- Vemos que ambas tablas tienen una columna nombrada **Unnamed: 0** que se puede eliminar sin afectar a los datos, además no siguen correctamente un orden.

En **netflix_originals.csv**:
- Tenemos 6 columnas que podrían ser útiles, **Título, Género,Premiere,Runtime,IMDB Score y Language.**
- Habría que preguntar si Runtime está en minutos en **netflix_originals.csv**.
- Vemos que no hay ninguna columna que tenga valores nulos en **netflix_originals.csv**.

En **netflix_titles.csv**:
- La primera columna sobra, unnamed 0, mirando el csv me hace sospechar de que salta numeraciones, lo que lo haría un index poco robusto
- Show_id se puede usar de índice al no tener valores duplicados
- Vemos que tenemos valores nulos en:
    - Director: **2634** valores nulos
    - Cast: **825** valores nulos
    - Country: **831** valores nulos
    - Date_added: **10** valores nulos
    - Duration: **4813** valores nulos
- Duración tiene valores distintos (minutos, temporadas...) lo que significa que la primera tabla sera tipo Movies todas.

En **netflix_originals.csv**
- Title: object
- Genre: object - 
- Premiere: object - Debería ser formato DATETIME
- Runtime: int - es correcto
- IMDB Score: float - es  correcto
- Language: object - es correcto

En **netflix_titles.csv**
- show_id: object 
- type: object - correcto, aquí seran todas **Movie** del primer CSV
- title: object - correcto, está en ambas ambos csv **Title**
- [con nulos] director: object - correcto
- [con nulos ]cast: object - correcto
- [con nulos ]country: object - correcto
- [con nulos] date_added: object - Por como está escrito es correcto., debería ser DATETIME
    - Aquí podemos poner la **Premiere** de la primera tabla, al ser un netflix original se estrena el mismo día que se añade a la plataforma.
- release_year: int - Es correcto al ser solo el año, podremos extraerlo para los originals
- rating: object - Correcto 
- [con nulos] duration: object - Correcto es donde va **Runtime**
- listed_in: object - Correcto, es el **Genre** del primer CSV
- description: object - Correcto

del Primer CSV, podemos unir al segundo:
- **Title**: En **title**
- **Genre**: En **listed_in**
- **Premiere**: En **date_added**
- **Runtime**: En **duration**

De forma individual, no hay títulos duplicados

En **netflix_originals.csv**
- Hay **513** títulos repartidos entre **106** categorías distintas

En **netflix_titles.csv**
- Hay **8807** títulos repartidos entre **514** categorías distintas



# Parte 2

In [45]:
netflix_ogs = pd.read_csv("datos/netflix_originals.csv")
netflix_titles = pd.read_csv("datos/netflix_titles.csv")

#originales = pd.read_csv("datos/netflix_originals.csv",usecols=["Title"])
netflix_originales = netflix_ogs.iloc[:,1:7]
netflix_originales = netflix_originales.rename(columns={"Title":"title","Genre":"genre","Premiere":"premiere","Runtime":"runtime"})
netflix_originales

title                  genre  \
0                                Enter the Anime            Documentary   
1                                    Dark Forces               Thriller   
2                                        The App  Science fiction/Drama   
3                                 The Open House        Horror thriller   
4                                    Kaali Khuhi                Mystery   
..                                           ...                    ...   
508                       Cuba and the Cameraman            Documentary   
509                       Dancing with the Birds            Documentary   
510  Winter on Fire: Ukraine's Fight for Freedom            Documentary   
511                      Springsteen on Broadway           One-man show   
512     David Attenborough: A Life on Our Planet            Documentary   

              premiere  runtime  IMDB Score                  Language  
0       August 5, 2019       58         2.5          English/Japanese  
1      August 21, 2020       81         2.6                   Spanish  
2    December 26, 2019       79         2.6                   Italian  
3     January 19, 2018       94         3.2                   English  
4     October 30, 2020       90         3.4                     Hindi  
..                 ...      ...         ...                       ...  
508  November 24, 2017      114         8.3                   English  
509   October 23, 2019       51         8.3                   English  
510    October 9, 2015       91         8.4  English/Ukranian/Russian  
511  December 16, 2018      153         8.5                   English  
512    October 4, 2020       83         9.0                   English  

[513 rows x 6 columns]

In [34]:
netflix_originales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 513 entries, 0 to 512
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   title       513 non-null    object 
 1   listed_in   513 non-null    object 
 2   date_added  513 non-null    object 
 3   duration    513 non-null    int64  
 4   IMDB Score  513 non-null    float64
 5   Language    513 non-null    object 
dtypes: float64(1), int64(1), object(4)
memory usage: 24.2+ KB


In [37]:
netflix_originales = netflix_originales.astype({"duration": "object"})
netflix_originales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 513 entries, 0 to 512
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   title       513 non-null    object 
 1   listed_in   513 non-null    object 
 2   date_added  513 non-null    object 
 3   duration    513 non-null    object 
 4   IMDB Score  513 non-null    float64
 5   Language    513 non-null    object 
dtypes: float64(1), object(5)
memory usage: 24.2+ KB


In [17]:
#Añadir las columnas que faltan

In [46]:
netflix_titles = pd.read_csv("datos/netflix_titles.csv")
netflix_titulos = netflix_titles.iloc[:,[1,2,3,4,6,7,8,10,11]]
netflix_titulos


show_id     type                  title         director        country  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson  United States   
1         s2  TV Show          Blood & Water              NaN   South Africa   
2         s3  TV Show              Ganglands  Julien Leclercq            NaN   
3         s4  TV Show  Jailbirds New Orleans              NaN            NaN   
4         s5  TV Show           Kota Factory              NaN          India   
...      ...      ...                    ...              ...            ...   
8802   s8803    Movie                 Zodiac    David Fincher  United States   
8803   s8804  TV Show            Zombie Dumb              NaN            NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer  United States   
8805   s8806    Movie                   Zoom     Peter Hewitt  United States   
8806   s8807    Movie                 Zubaan      Mozez Singh          India   

              date_added  release_year   duration  \
0     September 25, 2021          2020     90 min   
1     September 24, 2021          2021  2 Seasons   
2     September 24, 2021          2021        NaN   
3     September 24, 2021          2021        NaN   
4     September 24, 2021          2021  2 Seasons   
...                  ...           ...        ...   
8802   November 20, 2019          2007    158 min   
8803        July 1, 2019          2018  2 Seasons   
8804    November 1, 2019          2009        NaN   
8805    January 11, 2020          2006        NaN   
8806       March 2, 2019          2015    111 min   

                                              listed_in  
0                                         Documentaries  
1       International TV Shows, TV Dramas, TV Mysteries  
2     Crime TV Shows, International TV Shows, TV Act...  
3                                Docuseries, Reality TV  
4     International TV Shows, Romantic TV Shows, TV ...  
...                                                 ...  
8802                     Cult Movies, Dramas, Thrillers  
8803             Kids' TV, Korean TV Shows, TV Comedies  
8804                            Comedies, Horror Movies  
8805                 Children & Family Movies, Comedies  
8806     Dramas, International Movies, Music & Musicals  

[8807 rows x 9 columns]

In [51]:
ver_mismo = netflix_titulos.merge(netflix_originales, on=["title"])
ver_mismo

show_id   type                                             title  \
0        s1  Movie                              Dick Johnson Is Dead   
1      s142  Movie                                        Extraction   
2      s625  Movie                                        Prime Time   
3      s835  Movie                                      Blue Miracle   
4      s837  Movie                                         Ghost Lab   
..      ...    ...                                               ...   
508   s5902  Movie                                               Tig   
509   s5907  Movie                       What Happened, Miss Simone?   
510   s5911  Movie                                  Hot Girls Wanted   
511   s5913  Movie  The Other One: The Long Strange Trip of Bob Weir   
512   s5917  Movie                                        My Own Man   

                          director                                country  \
0                  Kirsten Johnson                          United States   
1                 Steven C. Miller  United States, United Kingdom, Canada   
2                     Jakub Piątek                                    NaN   
3                   Julio Quintana                          United States   
4              Paween Purijitpanya                               Thailand   
..                             ...                                    ...   
508  Kristina Goolsby, Ashley York                          United States   
509                     Liz Garbus                          United States   
510       Jill Bauer, Ronna Gradus                          United States   
511                    Mike Fleiss                          United States   
512                David Sampliner                          United States   

             date_added  release_year duration  \
0    September 25, 2021          2020   90 min   
1     September 1, 2021          2015   82 min   
2         June 30, 2021          2021      NaN   
3          May 27, 2021          2021      NaN   
4          May 27, 2021          2021  118 min   
..                  ...           ...      ...   
508       July 17, 2015          2015      NaN   
509       June 26, 2015          2015      NaN   
510        May 29, 2015          2015   83 min   
511        May 22, 2015          2015   84 min   
512       March 6, 2015          2015   82 min   

                                             listed_in        genre  \
0                                        Documentaries  Documentary   
1                                   Action & Adventure       Action   
2              Dramas, International Movies, Thrillers     Thriller   
3    Children & Family Movies, Dramas, Faith & Spir...        Drama   
4       Horror Movies, International Movies, Thrillers       Horror   
..                                                 ...          ...   
508                     Documentaries, Stand-Up Comedy  Documentary   
509                    Documentaries, Music & Musicals  Documentary   
510                                      Documentaries  Documentary   
511                    Documentaries, Music & Musicals  Documentary   
512                                      Documentaries  Documentary   

              premiere  runtime  IMDB Score Language  
0      October 2, 2020       90         7.5  English  
1       April 24, 2020      117         6.7  English  
2       April 14, 2021       91         5.7   Polish  
3         May 27, 2021       95         6.7  English  
4         May 26, 2021      117         5.2     Thai  
..                 ...      ...         ...      ...  
508      July 17, 2015       80         7.4  English  
509      June 26, 2015       84         7.6  English  
510       May 29, 2015       84         6.1  English  
511       May 22, 2015       83         7.3  English  
512  December 13, 2014       81         6.4  English  

[513 rows x 14 columns]

In [48]:
# Encontrar duplicados entre tablas 
unir_info = netflix_titulos.merge(netflix_originales, on=["title"],how="left")
unir_info

show_id     type                  title         director        country  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson  United States   
1         s2  TV Show          Blood & Water              NaN   South Africa   
2         s3  TV Show              Ganglands  Julien Leclercq            NaN   
3         s4  TV Show  Jailbirds New Orleans              NaN            NaN   
4         s5  TV Show           Kota Factory              NaN          India   
...      ...      ...                    ...              ...            ...   
8802   s8803    Movie                 Zodiac    David Fincher  United States   
8803   s8804  TV Show            Zombie Dumb              NaN            NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer  United States   
8805   s8806    Movie                   Zoom     Peter Hewitt  United States   
8806   s8807    Movie                 Zubaan      Mozez Singh          India   

              date_added  release_year   duration  \
0     September 25, 2021          2020     90 min   
1     September 24, 2021          2021  2 Seasons   
2     September 24, 2021          2021        NaN   
3     September 24, 2021          2021        NaN   
4     September 24, 2021          2021  2 Seasons   
...                  ...           ...        ...   
8802   November 20, 2019          2007    158 min   
8803        July 1, 2019          2018  2 Seasons   
8804    November 1, 2019          2009        NaN   
8805    January 11, 2020          2006        NaN   
8806       March 2, 2019          2015    111 min   

                                              listed_in        genre  \
0                                         Documentaries  Documentary   
1       International TV Shows, TV Dramas, TV Mysteries          NaN   
2     Crime TV Shows, International TV Shows, TV Act...          NaN   
3                                Docuseries, Reality TV          NaN   
4     International TV Shows, Romantic TV Shows, TV ...          NaN   
...                                                 ...          ...   
8802                     Cult Movies, Dramas, Thrillers          NaN   
8803             Kids' TV, Korean TV Shows, TV Comedies          NaN   
8804                            Comedies, Horror Movies          NaN   
8805                 Children & Family Movies, Comedies          NaN   
8806     Dramas, International Movies, Music & Musicals          NaN   

             premiere  runtime  IMDB Score Language  
0     October 2, 2020     90.0         7.5  English  
1                 NaN      NaN         NaN      NaN  
2                 NaN      NaN         NaN      NaN  
3                 NaN      NaN         NaN      NaN  
4                 NaN      NaN         NaN      NaN  
...               ...      ...         ...      ...  
8802              NaN      NaN         NaN      NaN  
8803              NaN      NaN         NaN      NaN  
8804              NaN      NaN         NaN      NaN  
8805              NaN      NaN         NaN      NaN  
8806              NaN      NaN         NaN      NaN  

[8807 rows x 14 columns]

In [49]:
#Exportamos a csv
#unir_info.to_csv("datos/netflix_union.csv")

# Explicación de la unión
# Para la unión
- Eliminamos Unnamed 0 de ambas filas
De **netflix_originals.csv**
- Nos quedamos con todas las columnas salvo unnamed0

De **netflix_titles.csv**
- No vamos a necesitar:
    - description: no aporta nada de valor a nuestro análisis.
    - rating: Tampoco nos aporta nada relevante.
    - cast: No aplica en el análisis que queremos hacer
- Nos quedamos con:
    - show_id: Que nos ayudará a separar los originales de netflix del resto
    - type: donde añadiremos los originales a "Movie"
    - title: Básico
    - director: Necesario en laboratorios posteriores
    - country: Necesario en laboratorios posteriores
    - date_added: Necesario para labs posteriores
    - release year: Para saber en que años hemos sacado más películas
    - duration: Algo para todos
    - listed_in: Las Categorías disponibles

Para realizar la unión primero probe con un **concat** a hacer esto:
- Title --> title: Son lo mismo en ambas tablas.
- Genre --> listed_in
- Premiere --> date_added
- Runtime --> duration
- IMDB Score: Será nueva columna
- Language: Será nueva columna

Lo que pasa, es que viendo con **merge** me di cuenta de que las películas se encuentran duplicadas, y que tenemos alguna información que está con dos valores distintos para la misma película, así que nos quedamos con las tablas como están.

La tabla final se verá así

**Show_id | type | title | title | director | country | date_added | release_year | duration | listed_in | genre | premiere | runtime | IMDB Score | Language**

Puntos importantes y actualizaciones de Conclusión:
- date added es cuando lo añadieron, pero si es un netflix original, deberían haberlo sacado en esa misma fecha **Los datos no dicen eso**

- duration y runtime deberían ser lo mismo, algunos tienen datos en ambas con valores distintos

- listed_in y genre son columnas distintas, es como categoría general y categoría exacta.

- type, efectivamente es "Movie"



### Parte 3: Uso de `loc` e `iloc`

Usando el DataFrame creado en la Parte 2: 





1. **Filtrado de datos con `loc`:**

   - Selecciona todas las filas donde el tipo de contenido sea “Movie”.



In [54]:
tabla_unida = unir_info
tabla_unida.loc[tabla_unida["type"]=="Movie"]

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                country          date_added  \
0                                         United States  September 25, 2021   
6                                                   NaN  September 24, 2021   
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year duration                                         listed_in  \
0             2020   90 min                                     Documentaries   
6             2021      NaN                          Children & Family Movies   
7             1993  125 min  Dramas, Independent Movies, International Movies   
9             2021  104 min                                  Comedies, Dramas   
12            2021  127 min                      Dramas, International Movies   
...            ...      ...                                               ...   
8801          2015      NaN           Dramas, International Movies, Thrillers   
8802          2007  158 min                    Cult Movies, Dramas, Thrillers   
8804          2009      NaN                           Comedies, Horror Movies   
8805          2006      NaN                Children & Family Movies, Comedies   
8806          2015  111 min    Dramas, International Movies, Music & Musicals   

            genre         premiere  runtime  IMDB Score Language  
0     Documentary  October 2, 2020     90.0         7.5  English  
6             NaN              NaN      NaN         NaN      NaN  
7             NaN              NaN      NaN         NaN      NaN  
9             NaN              NaN      NaN         NaN      NaN  
12            NaN              NaN      NaN         NaN      NaN  
...           ...              ...      ...         ...      ...  
8801          NaN              NaN      NaN         NaN      NaN  
8802          NaN              NaN      NaN         NaN      NaN  
8804          NaN              NaN      NaN         NaN      NaN  
8805          NaN              NaN      NaN         NaN      NaN  
8806          NaN              NaN      NaN         NaN      NaN  

[6131 rows x 14 columns]

   - Muestra solo los títulos (title) y la duración (duration) de todos los contenidos que sean de tipo “TV Show”.


In [57]:
tabla_unida.loc[tabla_unida["type"]== "TV Show",["title","duration"]]

title   duration
1             Blood & Water  2 Seasons
2                 Ganglands        NaN
3     Jailbirds New Orleans        NaN
4              Kota Factory  2 Seasons
5             Midnight Mass        NaN
...                     ...        ...
8795        Yu-Gi-Oh! Arc-V  2 Seasons
8796             Yunus Emre  2 Seasons
8797              Zak Storm        NaN
8800     Zindagi Gulzar Hai        NaN
8803            Zombie Dumb  2 Seasons

[2676 rows x 2 columns]


   - Selecciona todas las filas donde el país sea “United States”.



In [58]:
tabla_unida[tabla_unida["country"]== "United States"]

show_id     type                 title         director        country  \
0         s1    Movie  Dick Johnson Is Dead  Kirsten Johnson  United States   
9        s10    Movie          The Starling   Theodore Melfi  United States   
15       s16  TV Show     Dear White People              NaN  United States   
27       s28    Movie             Grown Ups     Dennis Dugan  United States   
28       s29    Movie            Dark Skies    Scott Stewart  United States   
...      ...      ...                   ...              ...            ...   
8791   s8792    Movie           Young Adult    Jason Reitman  United States   
8793   s8794    Movie  Yours, Mine and Ours     Raja Gosnell  United States   
8802   s8803    Movie                Zodiac    David Fincher  United States   
8804   s8805    Movie            Zombieland  Ruben Fleischer  United States   
8805   s8806    Movie                  Zoom     Peter Hewitt  United States   

              date_added  release_year duration  \
0     September 25, 2021          2020   90 min   
9     September 24, 2021          2021  104 min   
15    September 22, 2021          2021      NaN   
27    September 20, 2021          2010      NaN   
28    September 19, 2021          2013      NaN   
...                  ...           ...      ...   
8791   November 20, 2019          2011      NaN   
8793   November 20, 2019          2005      NaN   
8802   November 20, 2019          2007  158 min   
8804    November 1, 2019          2009      NaN   
8805    January 11, 2020          2006      NaN   

                                 listed_in        genre         premiere  \
0                            Documentaries  Documentary  October 2, 2020   
9                         Comedies, Dramas          NaN              NaN   
15                  TV Comedies, TV Dramas          NaN              NaN   
27                                Comedies          NaN              NaN   
28         Horror Movies, Sci-Fi & Fantasy          NaN              NaN   
...                                    ...          ...              ...   
8791  Comedies, Dramas, Independent Movies          NaN              NaN   
8793    Children & Family Movies, Comedies          NaN              NaN   
8802        Cult Movies, Dramas, Thrillers          NaN              NaN   
8804               Comedies, Horror Movies          NaN              NaN   
8805    Children & Family Movies, Comedies          NaN              NaN   

      runtime  IMDB Score Language  
0        90.0         7.5  English  
9         NaN         NaN      NaN  
15        NaN         NaN      NaN  
27        NaN         NaN      NaN  
28        NaN         NaN      NaN  
...       ...         ...      ...  
8791      NaN         NaN      NaN  
8793      NaN         NaN      NaN  
8802      NaN         NaN      NaN  
8804      NaN         NaN      NaN  
8805      NaN         NaN      NaN  

[2818 rows x 14 columns]

   - Muestra todas las películas (type = “Movie”) y selecciona las columnas title y director.



In [63]:
tabla_unida.loc[tabla_unida["type"]== "Movie",["title","director"]]

title                       director
0                 Dick Johnson Is Dead                Kirsten Johnson
6     My Little Pony: A New Generation  Robert Cullen, José Luis Ucha
7                              Sankofa                   Haile Gerima
9                         The Starling                 Theodore Melfi
12                        Je Suis Karl            Christian Schwochow
...                                ...                            ...
8801                           Zinzana                Majid Al Ansari
8802                            Zodiac                  David Fincher
8804                        Zombieland                Ruben Fleischer
8805                              Zoom                   Peter Hewitt
8806                            Zubaan                    Mozez Singh

[6131 rows x 2 columns]


   - Selecciona los títulos (title) y géneros (listed_in) de todas las producciones lanzadas en 2018.


In [64]:
tabla_unida.loc[tabla_unida["release_year"] == 2018,["title","listed_in"]]

title  \
21                               Resurrection: Ertugrul   
37                                          Angry Birds   
73                                         King of Boys   
94                                            Show Dogs   
96    If I Leave Here Tomorrow: A Film About Lynyrd ...   
...                                                 ...   
8732                                      White Chamber   
8743                                           Wildlife   
8774                                          يوم الدين   
8775                                    Yeh Meri Family   
8803                                        Zombie Dumb   

                                              listed_in  
21    International TV Shows, TV Action & Adventure,...  
37                                Kids' TV, TV Comedies  
73                         Dramas, International Movies  
94                   Children & Family Movies, Comedies  
96                      Documentaries, Music & Musicals  
...                                                 ...  
8732                        Sci-Fi & Fantasy, Thrillers  
8743                         Dramas, Independent Movies  
8774   Dramas, Independent Movies, International Movies  
8775                International TV Shows, TV Comedies  
8803             Kids' TV, Korean TV Shows, TV Comedies  

[1147 rows x 2 columns]


   - Selecciona las filas donde la columna director esté vacía y muestra solo los títulos (title).

In [69]:
tabla_unida.loc[tabla_unida["director"].isnull(),["title"]]

title
1                           Blood & Water
3                   Jailbirds New Orleans
4                            Kota Factory
10    Vendetta: Truth, Lies and The Mafia
14        Crime Stories: India Detectives
...                                   ...
8795                      Yu-Gi-Oh! Arc-V
8796                           Yunus Emre
8797                            Zak Storm
8800                   Zindagi Gulzar Hai
8803                          Zombie Dumb

[2634 rows x 1 columns]

2. **Acceso a datos específicos con `iloc`:**

   - Selecciona las primeras 5 filas del conjunto de datos y muestra solo las columnas title, director y country.



   - Muestra las últimas 5 filas del conjunto de datos y selecciona las columnas title y description.



   - Selecciona las primeras 10 filas del dataset y muestra las columnas desde la segunda hasta la quinta (incluidas). 



   - Muestra las últimas 7 filas del dataset y selecciona las columnas title, director, y country usando índices posicionales.



   - Selecciona las filas desde la 15 hasta la 25 (incluidas) y muestra las primeras 4 columnas.



   - Muestra las filas 20 a 30 y solo las columnas title, release_year, y rating usando índices posicionales.